spi/src/main/java/org/jboss/pnc/spi/datastore/predicates/ArtifactPredicates.java

```diff
@@ -79,21 +78,27 @@ public static Predicate<Artifact> withDependantBuildRecordId(Base32LongID buildR
 
     public static Predicate<Artifact> withDeliveredInProductMilestone(Integer productMilestoneId) {
         return (root, query, cb) -> {
-            Root<DeliverableArtifact> deliverableArtifact = query.from(DeliverableArtifact.class);
-            Join<DeliverableArtifact, DeliverableAnalyzerReport> report = deliverableArtifact
-                    .join(DeliverableArtifact_.report);
-            Join<DeliverableAnalyzerReport, DeliverableAnalyzerOperation> operation = report
-                    .join(DeliverableAnalyzerReport_.operation);
-
-            return cb.and(
-                    DeliverableAnalyzerReportPredicates.notFromDeletedAnalysis(cb, report),
-                    DeliverableAnalyzerReportPredicates.notFromScratchAnalysis(cb, report),
-                    cb.equal(
-                            operation.get(DeliverableAnalyzerOperation_.productMilestone).get(ProductMilestone_.id),
-                            productMilestoneId),
-                    cb.equal(
-                            deliverableArtifact.get(DeliverableArtifact_.artifact).get(Artifact_.id),
-                            root.get(Artifact_.id)));
+            Subquery<Artifact> subquery = query.subquery(Artifact.class);
+
+            Root<DeliverableAnalyzerOperation> deliverableAnalyzerOperations = subquery
+                    .from(DeliverableAnalyzerOperation.class);
+            Join<DeliverableAnalyzerOperation, ProductMilestone> productMilestones = deliverableAnalyzerOperations
+                    .join(DeliverableAnalyzerOperation_.productMilestone);
+            Join<DeliverableAnalyzerOperation, DeliverableAnalyzerReport> deliverableAnalyzerReports = deliverableAnalyzerOperations
+                    .join(DeliverableAnalyzerOperation_.report);
+            Join<DeliverableArtifact, Artifact> artifacts = deliverableAnalyzerReports
+                    .join(DeliverableAnalyzerReport_.artifacts)
+                    .join(DeliverableArtifact_.artifact);
+
+            subquery.select(artifacts);
+            subquery.where(
+                    cb.and(
+                            DeliverableAnalyzerReportPredicates.notFromDeletedAnalysis(cb, deliverableAnalyzerReports),
+                            DeliverableAnalyzerReportPredicates.notFromScratchAnalysis(cb, deliverableAnalyzerReports),
+                            cb.equal(productMilestones.get(ProductMilestone_.id), productMilestoneId)));
+            subquery.groupBy(artifacts);
```

Comment:
Do you need `groupBy()` if you're using `root.in()`? Having duplicate date with `in()` shouldn't matter right?

Reply:
no, it's a left-over; removed